### PR TITLE
ENH: performance boost by vectorizing call to funeval

### PR DESCRIPTION
**Pull request by sglyon**
the BasisMatrices routines are unfortunately not very efficient when evaluated at a single point.

This pull request tries to get around this by vectorizing the call to `funeval` in the innermost part of the `s_wise_max` set of routines.

The speedup was considerable. When solving a cdp with VFI at a tol of 1e-3, the monetary policy example went from 25 to 4 seconds and the optimal growth example went from 70 seconds to 8 seconds. 

Another (related) change is using type parameters for all the function fields of ContinuousDP. `Function` is an abstract type in Julia 0.6, and all functions are concrete subtypes of Function. Thus, we if let ContinuousDP parameterize on these concrete types, all uses of `cdp.(f|g|x_lb|x_ub)` will be type-inferrable by the compiler.

- - -

**EDIT**

I still believe there are many performance optimizations on the table, but this is a first step.

I really should have focused on getting the API stable, but I couldn't resist. 

- - -

### CODE REVIEW

**Comment by oyamad** [comment]
(Added some commits and rebased.)

Many thanks, this is great!

* [Before](http://nbviewer.jupyter.org/github/QuantEcon/ContinuousDPs.jl/blob/a59c262c869e22d14ad2091c4fd6b81212f3e62c/examples/cdp_ex_optgrowth_jl.ipynb)
* [After](http://nbviewer.jupyter.org/github/QuantEcon/ContinuousDPs.jl/blob/f83f1df6db5d4f0359fbc691900dfec9779c8b6e/examples/cdp_ex_optgrowth_jl.ipynb)

**Comment by sglyon** [comment]
Looks good to me. Happy to see that performance boost!

File: src/cdp.jl [hunk 1]
```diff
@@ -145,17 +148,15 @@ end
...
 function _s_wise_max(cdp::ContinuousDP, s, C)
+    sp = Array{Float64}(size(cdp.shocks, 1), length(s))
     function objective(x)
-        out = 0.
-        t = Base.tail(indices(cdp.shocks))
-        for (i, w) in enumerate(cdp.weights)
-            e = cdp.shocks[(i, t...)...]
-            out += w * funeval(C, cdp.interp.basis, cdp.g(s, x, e))
+        for i in 1:size(sp, 1)
+            sp[i, :] = cdp.g(s, x, cdp.shocks[i, :])
```
Comment (oyamad) [inline src/cdp.jl]: I was supposing the dimension of `cdp.shocks` may be larger than 2 (for example, the realization of shocks may be a matrix), but let's just assume that the shocks are scalar or vector (1de101f447c90c819bd7ec4a240a8cadbf5331b9).

Comment (sglyon) [reply src/cdp.jl]: Ahh ok, good point. I think matrix should be sufficiently general as higher dimension arrays can always be reshaped into a matrix and will naturally be stacked as columns due to the column major array ordering

**Pull request merged**